src/Components/Endpoints/src/Rendering/SSRRenderModeBoundary.cs

```diff
@@ -106,15 +108,20 @@ private void Prerender(RenderTreeBuilder builder)
             var serverComponentSerializer = httpContext.RequestServices.GetRequiredService<ServerComponentSerializer>();
 
             var invocationId = EndpointHtmlRenderer.GetOrCreateInvocationId(httpContext);
-            serverMarker = serverComponentSerializer.SerializeInvocation(invocationId, _componentType, parameters, _prerender);
+            serverMarker = serverComponentSerializer.SerializeInvocation(invocationId, _componentType, parameters, markerKey, _prerender);
         }
 
         WebAssemblyComponentMarker? webAssemblyMarker = null;
         if (_renderMode is WebAssemblyRenderMode or AutoRenderMode)
         {
-            webAssemblyMarker = WebAssemblyComponentSerializer.SerializeInvocation(_componentType, parameters, _prerender);
+            webAssemblyMarker = WebAssemblyComponentSerializer.SerializeInvocation(_componentType, parameters, markerKey, _prerender);
         }
 
         return (serverMarker, webAssemblyMarker);
     }
+
+    private string GenerateMarkerKey(int sequence, object? key)
+    {
+        return HashCode.Combine(_componentType.FullName, sequence, key).ToString("x", CultureInfo.InvariantCulture);
```

Comment:
Have we measured the chances of collision here? `GetHashCode` is not collision resistant.

Reply:
I left a comment about that [here](https://github.com/dotnet/aspnetcore/pull/49238#discussion_r1264762420). I'm definitely open to improving this, because a collision is definitely possible.

We could reduce the possibility of a collision by generating a string that separately includes the following:
* Sequence number
* Key hash code
* Type name hash code

We could even go further and obtain a SHA1 hash for the type name.

Reply:
I think we should still send the type name + sequence number without hashing. We can hash the key if there is any, but if we are using this for comparison, it feels like we are shooting ourselves in the foot computing a "digest" for the key when we have all the data.

If we are concerned about the size of the TypeName, we can create a Map for each type that we render as a server component on the fly that guarantees their uniqueness, but I would start with just sending the data without losing any information.

Reply:
And is it fine that we might be sending the full name of a server-only type to the client? I wasn't sure if that would be considered a security risk. e.g., `Company.App.SomeComponent_AndByTheWayThePasswordIsFoo`.

Reply:
It's inside a data protected payload, isn't it?

Reply:
The key is not data protected. It wouldn't work to included it in the data protected payload because any change to parameter values would also change the payload's text. We could include it as a separate data protected payload, but then that's just a more expensive hash (and the key doesn't get round-tripped back to the server again anyway, it's only used by the JS code).

Reply:
@SteveSandersonMS I agree that from a security perspective, it's probably fine to hash the key and other data alongside it.

What do you think of the concern about hash collisions? Do you think it would be fine to alternatively send the full name of the component type to the client, and possibly suggest/require that the key implements `IFormattable`?

Reply:
I was thinking about hash collisions a bit yesterday.

It's not as risky as it would be if we needed every hash to be unique - if that were the case then a list of only ~4500 items would have a > 50% chance of a clash (via birthday paradox maths). We only need the keys to be unique enough that the edit operations that happen in practice don't run into clashes. If the edits are deletions/insertions, then we only need the *next/preceding* item not to have the same hash as the inserted/deleted item. So the probability of this being violated is simply 1/2^32. If the edit operation was randomly shuffling a list, then the risk is harder to compute (I lost interest after a few minutes...).

> alternatively send the full name of the component type to the client

I don't think that helps much since the sequence number almost always uniquely identifies the type anyway. In fact it always would if the code was compiled from `.razor`, since there's no way in Razor syntax to render a child whose type varies without the sequence number varying. The one case where this is not true is if you deploy a new version of your app in between the two requests - that's the one edge case I think it's worth retaining a hash of the type name for.

> possibly suggest/require that the key implements IFormattable?

Yeah, that is a good possible solution. We could say that we will *only* include the `@key` value in the marker if it implements `IFormattable` (and if it doesn't, that's not an error since it may be used for interactive rendering, but we simply ignore the key in that case).

The other good thing about this is it avoids false negatives. That is, if someone was using `@key=someModelObjectFromTheDatabase` which is completely valid and useful in interactive rendering:

 * If we use `(type, sequence, key?.GetHashCode())`, it will produce a different output on every request, perhaps causing us *never* to retain the interactive components
 * If we only use `(type, sequence, (key as IFormattable)?.ToString())` then we will err on the side of retaining interactive components, and `@key` becomes a way to optionally prevent unwanted retention in a predictable way. It also guarantees no false positive matches.

So, your `IFormattable` suggestion sounds like a better solution to me.

Reply:
> The one case where this is not true is if you deploy a new version of your app in between the two requests - that's the one edge case I think it's worth retaining a hash of the type name for.

Actually forget that - it's an irrelevant scenario. If there is a new version of the app between the two requests, it would only affect WebAssembly because that's the only kind of interactive component that could survive a server restart.

It *is* still relevant to avoid accidental matches based on `sequence` alone because the circuit would reject any incoming parameters for a componentId that doesn't match the component type encoded in the Data Protected payload[1]. So there'd be an exception if there's an accidental match for a different component type. And that could happen because when we merge the DOM trees, some parent component might have contained an `@if`/`@else` branch that outputs two identical DOM structures that place two different component types at the equivalent output location and happen to have the same `sequence` number at that point in rendering.

So, to make things much more robust in the default case where there's no `@key`, I still think we should be including some information based on the component type. We could use the full type name if we are OK with disclosing that. Or if not, we'd want some stable value derived from it. But then we have to decide if we rely on SSR client-server affinity or not. If we want to allow for a cluster of SSR servers without client affinity, we cannot use either:

 * A `Type`->number lookup table maintained in server memory
 * `type.FullName.GetHashCode()`, since `GetHashCode` isn't required to be stable across processes

Instead we'd need a stable, server-independent hash of the type name. For example the server could maintain a cache of `Type`->`SHA(type.FullName)`.

[1] We do verify and enforce that, right?

Reply:
Great analysis, @SteveSandersonMS!

> It is still relevant to avoid accidental matches based on sequence alone because the circuit would reject any incoming parameters for a `componentId` that doesn't match the component type encoded in the Data Protected payload... We do verify and enforce that, right?

Yes, if there is a type mismatch between the type in the data protected payload and the type of the component with the supplied ID, we reject the parameter update (see line 130 below).

https://github.com/dotnet/aspnetcore/blob/d3397906f11eb82a3b01780272f77d66eaf56765/src/Components/Server/src/Circuits/RemoteRenderer.cs#L123-L136

However, it's totally possible for the client to lie about the component ID if it happens to correspond to another root component of the same type. It would be tricky to come up with a mechanism to prevent that from happening, because it's the client who decides which SSR/interactive components match up anyway.

> `type.FullName.GetHashCode()`, since `GetHashCode` isn't required to be stable across processes

Interesting... I didn't know that was the case for strings. Good to know!

> For example the server could maintain a cache of Type->SHA(type.FullName)

This seems like the right approach to me.

Based on this discussion, it's looking like we'll generate a key with:
* The sequence number included directly
* A cached SHA for the component type name (maybe a truncated hash if we want to reduce payload size, as long as it doesn't make the risk of collision unacceptable)
* The formatted `@key`, if it's provided and implements `IFormattable`